src/Npgsql/PoolingDataSource.cs

```diff
@@ -146,8 +183,16 @@ class PoolingDataSource : NpgsqlDataSource
                 {
                     if (async)
                     {
+                        // If we couldn't reserve a connector enqueue as a blocked reader.
+                        // No sync TryRead on the channel (like multiplexing via TryGetIdleConnector) will succeed until we get a connector.
+                        // Given all connectors have been set as reserved, at some point:
+                        // 1. The last reserved connector will also become idle.
+                        // 2. Handed to a reserving waiter.
+                        // 3. Used for exlusive work and returned.
+                        // 4. Handed to the next waiter in queue (must be the one or one of the blocked reader(s) that failed to reserve).
+                        ReserveConnector();
```

Comment:
Maybe move the check for multiplexing out of the method, so we don't even call it if multiplexing isn't on (mostly for clarity, perf isn't that important here)